File: project/wavelet-dist/haar-threaded-xform/matrixType.m
//
//  matrixType.m
//  oc-matrix
//
//  Created by Daniel D. Beatty on 4/1/05.
//  Copyright 2005 __MyCompanyName__. All rights reserved.
//

#import "matrixType.h"
#import <stdlib.h>

@implementation matrixType
-(id) initWithRows: (int) rows columns:(int) cols
{
	m_rows = rows;
	m_cols = cols;
	[self allocateStorage];
	[self defineEpsilon];
}

-(id) initWithMatrixType:(matrixType *) rhs
{
}

-(void) copyElements: (matrixType *) source
{
	int i; 
	int j;
	double **temp = [source matrix];
	for ( i = 0; i < m_rows; i++)
		for ( j = 0; j < m_cols; j++) 
			m_data[i][j] = [source matrix][i][j];
	shiftX = [source shiftX];
	shiftY = [source shiftY];
	scaleX = [source scaleX];
	scaleY = [source scaleY];
}

-(void) allocateStorage
{
	int i;
	int k;
	if ( m_cols <= 0 || m_rows <= 0) 
	{
		m_data = NULL;
		return;
	}
	k =  m_rows * m_cols;
	m_data = malloc( sizeof(double*) * m_rows);
	m_data[0] = malloc ( sizeof(double) * k);
	for ( i = 1; i < m_rows; i++)
	{
		m_data[i] = m_data[i-1] + m_cols;
		printf ("memory location %d \n ", m_data[i] );
	}
}

-(void) allocateRows: (int) rows columns:(int)cols
{
	m_rows = rows;
	m_cols = cols;
	[self allocateStorage];
}

-(void) deleteStorage
{
	if (m_data != NULL) {
		free(m_data[0]);
		free(m_data);
		m_data = NULL;
	}
}

-(void) dealloc 
{
	[self deleteStorage];
}

-(void) zeroize
{
    int i, j;
    for ( i = 0; i < m_rows; i++)
        for ( j = 0; j < m_cols; j++)
            m_data[i][j] = 0;

}

-(void) defineEpsilon
{
	epsilon = 1/pow(2.0, 32.0);
}

-(BOOL) checkRows: (int) rows columns:(int)cols
{
	return ((rows>=0)&&(rows<m_rows))&&((cols>=0)&&(cols<m_cols)); 
}


-(int) m_cols 
{
	return m_cols;
}

-(int) m_rows 
{
	return m_rows;
}


-(void) addMatrixEqual: (matrixType *) rhs
{
	int i, j;
	[self checkRows:[rhs m_rows] columns:[rhs m_cols]];
	double **rhs_m_data = [rhs matrix];
	for (i=0; i< m_rows; i++)
		for ( j=0; j<m_cols; j++)
			m_data[i][j] += rhs_m_data[i][j];
}

-(void) subMatrixEqual: (matrixType *) rhs
{
	int i, j;
	[self checkRows:[rhs m_rows] columns:[rhs m_cols]];	
	double **rhs_m_data = [rhs matrix];
	for (i=0; i<m_rows; i++)
		for (j=0; j<m_cols; j++)
			m_data[i][j] -= rhs_m_data[i][j];
	
}


-(void) multMatrixEqual: (matrixType *) rhs
{
	int i, j, k;
	if ( m_cols == [rhs m_rows]) 
	{
		double total;
		int rhsCol = [ rhs m_rows];
		double **rhs_m_data = [rhs matrix];
		matrixType *result= [[matrixType init] alloc];
		[result allocateRows:m_rows columns:rhsCol];
		for ( i=0; i< m_rows; i++)
			for ( j=0; j< rhsCol; j++) {
				total = 0;
				for ( k=0; k< m_cols; k++)
					total += m_data[i][k]*rhs_m_data[k][j];
			//[result matrix][i][j] = total;
			}
	}
}

-(void) sparseMatrixMultiplyEqual: (matrixType *) rhs
{
}

-(void) multEqual: (double) rhs
{
	int i, j;
	for ( i=0; i<m_rows; i++)
		for ( j=0; j<m_cols; j++)
			m_data[i][j] *= rhs;
	
}

-(void) divEqual: (double) rhs
{
	int i, j;
	double temp = 1/rhs;
	for ( i=0; i<m_rows; i++)
		for ( j=0; j<m_cols; j++)
			m_data[i][j] *= temp;
	
}

-(void) addEqual: (double) rhs
{
	int i , j;
	for ( i=0; i<m_rows; i++)
		for ( j=0; j<m_cols; j++)
			m_data[i][j] += rhs;
}

-(void) subEqual: (double) rhs
{
	int i, j;
	for ( i=0; i<m_rows; i++)
		for ( j=0; j<m_cols; j++)
			m_data[i][j] -= rhs;
}

-(double) getEnergy 
{
	int i, j;
    double sumcol = 0;
    double sumrow = 0;
    for ( i = 0; i < m_rows; i++) {
        sumcol = 0;
        for ( j = 0 ; j < m_cols; j++)
        {
            sumcol += m_data[i][j] * m_data[i][j] / m_cols;
        }
        sumrow += sumcol / m_rows;
    }
    return sumrow;
}

-(double) getL1Energy 
{
    int i, j;
    double sumcol = 0;
    double sumrow = 0;
    for ( i = 0; i < m_rows; i++) {
        sumcol = 0;
        for ( j = 0 ; j < m_cols; j++)
        {
            sumcol += m_data[i][j];
        }
        sumrow += sumcol;
    }
    return sumrow;
}

-(double**) matrix
{
	return m_data;
}

-(double) shiftX
{
	return scaleY;
}

-(double) shiftY
{
	return scaleY;
}


-(double) scaleX
{
	return scaleY;
}

-(double) scaleY
{
	return scaleY;
}

-(double) getMin
{
	return min;
}

-(double) getMax
{
	return max;
}


-(double) getMaxWithIn: (int) Urows lRows:(int) lRows lCols:(int) lCols rCols:(int) rCols
{
	int i, j;
	int localMax = 0;
	for ( i = Urows; i < lRows; i++)
		for ( j = lCols; j < rCols; j++)
		{
			if (localMax <= m_data[i][j] ) localMax = m_data[i][j];
		}
			return localMax;
}

-(double) getMinWithIn: (int) Urows lRows:(int) lRows lCols:(int) lCols rCols:(int) rCols
{
	int i, j;
	int localMin = 0;
	for ( i = Urows; i < lRows; i++)
		for ( j = lCols; j < rCols; j++)
		{
			if (localMin <= m_data[i][j] ) localMin = m_data[i][j];
		}
			return localMin;
}

-(double) getEnergyWithIn: (int) Urows lRows:(int) lRows lCols:(int) lCols rCols:(int) rCols
{
	int i, j;
    double sumcol = 0;
    double sumrow = 0;
	double numRows = lRows - Urows;
	double numCols = lCols - rCols;
	double numColsInv = 1 /numCols ;
	double numRowsInv = 1/ numRows;
    for ( i = Urows; i < lRows; i++) {
        sumcol = 0;
        for ( j = lCols ; j < rCols; j++)
        {
            sumcol += m_data[i][j] * m_data[i][j] * numColsInv;
        }
        sumrow += sumcol * numRowsInv;
    }
    return sumrow;
}

-(double) getL1EnergyWithIn: (int) Urows lRows:(int) lRows lCols:(int) lCols rCols:(int) rCols
{
	int i, j;
    double sumcol = 0;
    double sumrow = 0;
	double numRows = lRows - Urows;
	double numCols = lCols - rCols;
	double numColsInv = 1 /numCols ;
	double numRowsInv = 1/ numRows;
    for ( i = Urows; i < lRows; i++) {
        sumcol = 0;
        for ( j = lCols ; j < rCols; j++)
        {
            sumcol += m_data[i][j]  * numColsInv;
        }
        sumrow += sumcol * numRowsInv;
    }
    return sumrow;
}


@end

File: trunk/original/wavelets/psijni/jnixform/matrixType.cpp
#include "matrixType.h"
//using namespace std;
#include <stddef.h>
#include <stdlib.h>
#include <unistd.h>
#include <iostream.h>
#include <math.h>

matrixType::matrixType () 
{
    m_rows = 0; 
    m_cols=0; 
    m_data = NULL;
    defineEpsilon ();
}

void matrixType::defineEpsilon ()
{
    epsilon = 1/pow(2.0, 32.0);
}

bool matrixType::copyElements (const matrixType& source)
    {
      for (int i=0; i<m_rows; i++)
            for (int j=0; j<m_cols; j++)
                m_data[i][j] = source.m_data[i][j];
        shiftX = source.shiftX;
        shiftY = source.shiftY;
        scaleX = source.scaleX;
        scaleY = source.scaleY;
        return true;
    }
    
    matrixType& matrixType::operator= (const matrixType& rhs)
    {
        if (this != &rhs) {
        if (m_rows != rhs.m_rows || m_cols != rhs.m_cols) {
                deleteStorage();
                m_rows = rhs.m_rows;
                m_cols = rhs.m_cols;
                allocateStorage();
            }
            copyElements(rhs);
            
        }
        return (*this);
    }
    
    
    bool matrixType::allocateStorage ()
    {
        if (m_rows <=0 || m_cols <=0) {
            m_data = NULL;
            return false;
        }

        m_data = new double*[m_rows];
        m_data[0] = new double[m_rows*m_cols];
        for (int i=1; i<m_rows; i++)
            m_data[i] = m_data[i-1]+m_cols;
        return true;
    }
    
    bool matrixType::allocateStorage (int rows, int cols)
    {
        m_rows = rows;
        m_cols = cols;
        return allocateStorage();
    }
    
    bool matrixType::deleteStorage ()
    {
        if (m_data != NULL) {
            delete []m_data[0];
            delete []m_data;
            m_data = NULL;
        }
        return true;
    }
    
bool matrixType::zeroize ()
{
    int i, j;
    for ( i = 0; i < m_rows; i++)
        for ( j = 0; j < m_cols; j++)
            m_data[i][j] = 0;
}
    
matrixType& matrixType::operator+=(const matrixType& rhs)
{
  checkSize(rhs.m_rows, rhs.m_cols);

  for (int i=0; i<m_rows; i++)
    for (int j=0; j<m_cols; j++)
      m_data[i][j] += rhs.m_data[i][j];
        

  return (*this);
}


/*
matrixType& matrixType::operator-=(const imageType& rhs)
{
  checkSize(rhs.m_rows, rhs.m_cols);

  for (int i=0; i<m_rows; i++)
    for (int j=0; j<m_cols; j++)
      m_data[i][j] -= rhs.m_data[i][j];

  return (*this);
}
*/

matrixType& matrixType::operator*=(const matrixType& rhs)
{
  if (m_cols != rhs.m_rows)
    error("operator*=: matrices not same inner size");

  double total;
  int rhsCol = rhs.m_rows;

  matrixType result(m_rows, rhsCol);
  for (int i=0; i< m_rows; i++)
    for (int j=0; j< rhsCol; j++) {
      total = 0;
      for (int k=0; k< m_cols; k++)
	total += m_data[i][k]*rhs.m_data[k][j];
      result.m_data[i][j] = total;
    }

  (*this) = result;
  return (*this);
}

matrixType& matrixType::sparsematrixmultiply (const matrixType &lhs, const matrixType &rhs)
{
    int k = lhs.m_cols;
    int i,j;
    int ival, jval;
    //matrixlist *sparsemat = new matrixlist[k];
    // Load sparse matrix 
    
    // Perform sparse multiply
    for ( i = 0; i < lhs.m_rows; i++)
        for ( j = 0; j < rhs.m_cols; j++)
        {
            
        }
}


matrixType& matrixType::operator*=(double rhs)
{
  for (int i=0; i<m_rows; i++)
    for (int j=0; j<m_cols; j++)
      m_data[i][j] *= rhs;

  return (*this);
}



matrixType& matrixType::operator/=(double rhs)
{
  for (int i=0; i<m_rows; i++)
    for (int j=0; j<m_cols; j++)
      m_data[i][j] /= rhs;

  return (*this);
}


matrixType& matrixType::operator+=(double rhs)
{
  for (int i=0; i<m_rows; i++)
    for (int j=0; j<m_cols; j++)
      m_data[i][j] += rhs;

  return (*this);
}



matrixType& matrixType::operator-=(double rhs)
{
  for (int i=0; i<m_rows; i++)
    for (int j=0; j<m_cols; j++)
      m_data[i][j] -= rhs;

  return (*this);
}


const matrixType operator*(const matrixType& lhs, const matrixType& rhs)
{
  matrixType u = lhs;
  u *= rhs;
  return u;
}


const matrixType operator*(const matrixType &lhs, double rhs)
{
  matrixType M = lhs;
  M *= rhs;
  return M;
}

const matrixType operator*(double lhs, const matrixType &rhs)
{
  matrixType M = rhs;
  M *= lhs;
  return M;
}


const matrixType operator+(const matrixType &lhs, const matrixType &rhs)
{
  matrixType M;
  M = lhs;
  M += rhs;
  return M;
}


const matrixType operator-(const matrixType &lhs, const matrixType &rhs)
{
  matrixType M;
  M = lhs;
  M += -1 * (rhs);
  return M;
}


bool matrixType::checkSize(int rows, int cols)
{
  return ((rows>=0)&&(rows<m_rows))&&((cols>=0)&&(cols<m_cols)); 
}

void matrixType::error(const char *errorMsg)
{
  cout << "ImageType: " << errorMsg << endl;
  exit(1);
}


void matrixType::GJ ()
{
    int i, j, l, k;
    double temp;
    for ( i = 0; i < m_rows; i++)
    {
        k = i;
        if  ((m_data[k][i] < epsilon ) && (m_data[k][i] > (-epsilon)))
        {
            while (( k < m_rows) && (m_data[k][i] < epsilon ) && (m_data[k][i] > (-epsilon)) )
        	k++;
            for ( j = i; j < m_cols; j++)
            {
                temp = m_data[k][j];
                m_data[k][j] = m_data[i][j];
                m_data[i][j] = m_data[k][j];
            }
        }
        temp = 1/ m_data[i][i];
        for ( k = i + 1; k < m_cols; k++)
            m_data[i][k] *= temp;
        m_data[i][i] = 1;
        for ( k = i + 1; k < m_rows; k++)
        {
            if ((m_data[k][i] < epsilon ) && (m_data[k][i] > (-epsilon)) )
            {
                temp = 1/m_data[k][i];
                for ( l = 0; l < m_cols; l++)
                {
                    m_data[k][l] *= -temp;
                    m_data[k][l] += m_data[i][l]; 
                }
                m_data[k][i] = 0;
                
            }
        }
        m_data[i][i] = 1.0;
    }
}

double matrixType::getEnergy ()
{
    int i, j;
    double sumcol = 0;
    double sumrow = 0;
    for ( i = 0; i < m_rows; i++) {
        sumcol = 0;
        for ( j = 0 ; j < m_cols; j++)
        {
            sumcol += m_data[i][j] * m_data[i][j] / m_cols;
        }
        sumrow += sumcol / m_rows;
    }
    return sumrow;
}

double matrixType::getL1Energy ()
{
    int i, j;
    double sumcol = 0;
    double sumrow = 0;
    for ( i = 0; i < m_rows; i++) {
        sumcol = 0;
        for ( j = 0 ; j < m_cols; j++)
        {
            sumcol += m_data[i][j];
        }
        sumrow += sumcol;
    }
    return sumrow;
}
